### ISSUE TITLE
Use LCS as a default compaction strategy for MVs

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#8639]
*Installation details*
HEAD: 84808399327c595ec7aa1cb7a7956d181baf9284

*Description*
Today we will always create an MV with a STCS (apparently we use regular tables' default).

However MV's updates are a read -> delete -> insert sequence.
Therefore it's sstables are always going to have a lot of tombstones and STCS is very bad for this kind of workload because it's terrible in getting rid of deleted data and those tombstones (after gc_grace_seconds). 

As a result the MV performance is going to degrade (likely significantly) as long as more WRITEs land at the base table hence triggering more MV updates.

There are 2 main standard ways of tackling this kind of situation:
1) STCS + periodic major compaction
2) LCS

I believe that using LCS for MVs as a default compaction strategy is a more practical solution.

**Comment by vladzcloudius** [comment]
@nyh @psarna @raphaelsc Please, comment.

**Comment by nyh** [comment]
@vladzcloudius you are correct that **after** gc_grace_seconds STCS finds it more difficult to get rid of tombstones. I was talking about what happens before that period, but you're right that in a setup running for months, it's probably more important what happens in the long run.

@vladzcloudius the problem with your suggestion is that making LCS the recommended (or even default) compaction strategy was suggested in the past, and rejected. In some scenarios (where we often write to not-very-recently-written rows), it has a much higher write amplification than STCS. You are suggesting that specifically for materialized views, the benefits (of faster removal of deleted data and tombstones) outweigh the negative aspects, but do we have a benchmark or a user report (or just some manual run you did) that demonstrates that? I'm not **against** changing the default compaction strategy - as you said users can easily override this default choice - but I think we need a bit more evidence that this is a good idea.

Also, changing from STCS to LCS won't suddenly make the tombstone problem disappear. All it will do is to reduce the scope of the problem from the amount of tombstones collected in time-between-major-compaction to gc_grace_seconds. The question is how long and how different these two time periods are. Which brings me to ask another question: What should the gc_grace_seconds be set to in view tables? Do people run repair on view tables?

**Comment by vladzcloudius** [comment]
> @vladzcloudius you are correct that **after** gc_grace_seconds STCS finds it more difficult to get rid of tombstones. I was talking about what happens before that period, but you're right that in a setup running for months, it's probably more important what happens in the long run.
> 
> @vladzcloudius the problem with your suggestion is that making LCS the recommended (or even default) compaction strategy was suggested in the past, and rejected. In some scenarios (where we often write to not-very-recently-written rows), it has a much higher write amplification than STCS. You are suggesting that specifically for materialized views, the benefits (of faster removal of deleted data and tombstones) outweigh the negative aspects, but do we have a benchmark or a user report (or just some manual run you did) that demonstrates that? I'm not **against** changing the default compaction strategy - as you said users can easily override this default choice - but I think we need a bit more evidence that this is a good idea.

@nyh 
I don't have benchmark results.
But we do have at least one user that switched to LCS for MVs recently and so far they haven't seen issues. 

More will follow I believe.

There are other accounts that run periodic majors on MVs and this is obviously a more time consuming option.

> 
> Also, changing from STCS to LCS won't suddenly make the tombstone problem disappear. All it will do is to reduce the scope of the problem from the amount of tombstones collected in time-between-major-compaction to gc_grace_seconds. The question is how long and how different these two time periods are. 

The idea of switching to LCS is to not have to run majors at all.

> Which brings me to ask another question: What should the gc_grace_seconds be set to in view tables? 

That's a very good question. If MV deletes are executed with CL=ALL then gc_grace_seconds may be set to 0 and no repair is needed in order to prevent the data resurrection.

> Do people run repair on view tables?

I believe they do.
But formally speaking MV requires a special kind of repair that would re-sync the content of the base table and the MV. That's because "regular" repairs of either the base tables or MV won't guarantee the consistent MV content in a general case.

**Comment by vladzcloudius** [comment]
> Also, changing from STCS to LCS won't suddenly make the tombstone problem disappear.

There are multiple "tombstones problems" when we talk about MVs and I agree that the main is that there are too many tombstones in the MV to begin with ;)
And you are right that LCS is not going to resolve that one.

However LCS is definitely a much better default choice than STCS.
And you are right that in a heavy-write-to-the-base case the disk can choke on the write amplification LCS is going to incur. In that case STCS + periodic majors would be the better choice.

But STCS (without majors) is a *bad* choice for MV no matter what use case that is.
Thus since we don't have an automatic-major option LCS is the best next default IMO.

**Comment by mykaul** [comment]
I think nowdays, we should use ICS.

**Comment by vladzcloudius** [comment]
> I think nowdays, we should use ICS.

Agree and we do.
But I believe there is a separate GH issue about ICS defaults that should be used for MVs.
In particular I suggested to always enable SAG (for MVs).
I believe @raphaelsc argued that a "new GC compaction" should be good enough to address the issue described above however AFAIR there was a series of edge case issues with the GC compaction that still didn't allow it to completely replace SAG.

The latter should be reviewed as well because it may be that the current state of ICS defaults and GC compaction in particular doesn't require any special adjustment of ICS for MV. 

I believe RaphaelC should be the best person to know what's the actual state of things.

Regardless I agree that this GH can be safely closed.

**Comment by nyh** [comment]
@vladzcloudius I think that ICS+SAG issue is https://github.com/scylladb/scylla-enterprise/issues/2332, which you opened. Could you please do me a favor and close this issue and instead open a new issue copying the main ideas from the enterprise issue (and close the enterprise issue)? Thanks.
